### PR TITLE
hookstate: refactor around EphemeralRunHook

### PR DESCRIPTION
**Pull request by mvo5**
This implements the feedback/ideas from https://github.com/snapcore/snapd/pull/9715#discussion_r533210795

### CODE REVIEW

File: overlord/hookstate/context.go [hunk 1]
```diff
@@ -74,6 +75,25 @@ func NewContext(task *state.Task, state *state.State, setup *HookSetup, handler
 	}, nil
 }
 
+func newEphemeralContext(st *state.State, setup *HookSetup, contextData map[string]interface{}) (*Context, error) {
```
Comment (pedronis) [inline overlord/hookstate/context.go:78]: I almost forgot but there's another place that makes an ephemeral context but with a provided cookie,  maybe for claritiy this should be called newEphemeralHookContextWithData

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/hookstate/context.go [hunk 2]
```diff
@@ -285,3 +283,30 @@ func (c *Context) ChangeID() string {
...
+// Logf logs to the context, either to the logger for ephemeral contexts
+// or the task log.
+//
+// Context must be locked.
+func (c *Context) Logf(fmt string, args ...interface{}) {
+	c.writing()
+	if c.IsEphemeral() {
+		logger.Noticef(fmt, args...)
+	} else {
+		c.task.Logf(fmt, args...)
```
Comment (stolowski) [inline overlord/hookstate/context.go:296]: Nice!

**Review by stolowski — APPROVED** [review]
LGTM, thank you!

**Pull request merged**